### PR TITLE
Add ability to perform an etcd on-demand snapshot via cli

### PR DESCRIPTION
**Pull request by briandowns**
Signed-off-by: Brian Downs <brian.downs@gmail.com>

#### Proposed Changes ####

This PR adds a new sub-command to the K3s command called `etcd-snapshot` and it takes a number of the standard K3s flags as well as specific flags like `name`, used to give the snapshot a specific name and `dir` used to tell K3s where to save the snapshot. The former use applicable in this use case as well as for cron scheduled snapshots and keeps in like with RKE features for snapshotting and will ultimately be pulled into RKE2. An on-demand snapshot will not effect snaptshot retention.

#### Verification ####

1. Start K3s using Etcd and wait for it to get to steady state. 
2. Run `k3s etcd-snapshot` to perform an etcd snapshot.
  A. Verify the snapshot has been performed successfully via log output `{"level":"info","msg":"saved","path":"/var/lib/rancher/k3s/server/db/snapshots/on-demand-<unix-timestamp>"}`
  B. Verify that the newly created snapshot is indeed at the expected location
3. Run `k3s etcd-snapshot --name=my_snapshot` to perform and etcd snapshot now.
  A. Perform the same verification steps from above.

#### Linked Issues ####

#2758

### CODE REVIEW

**Review by brandond — APPROVED** [review]

File: pkg/cli/cmds/server.go [hunk 1]
```diff
@@ -220,6 +220,7 @@ func NewServerCommand(action func(*cli.Context) error) cli.Command {
 				Name:        "etcd-snapshot-name",
 				Usage:       "(db) Set the name of etcd snapshots. Default: etcd-snapshot-<unix-timestamp>",
```
Comment (brandond) [inline pkg/cli/cmds/server.go:221]:
```suggestion
				Usage:       "(db) Set the base name of scheduled etcd snapshots. Default: etcd-snapshot-",
```

File: pkg/cli/cmds/etcd_snapshot.go [hunk 2]
```diff
@@ -24,19 +27,14 @@ func NewEtcdSnapshotCommand(action func(*cli.Context) error) cli.Command {
 				Destination: &ServerConfig.DataDir,
 			},
 			&cli.StringFlag{
-				Name:        "etcd-snapshot-name",
-				Usage:       "(db) Set the name of etcd snapshots. Default: etcd-snapshot-<unix-timestamp>",
+				Name:        "name",
+				Usage:       "(db) Set the name of the etcd on-demand snapshot. Default: on-demand-<unix-timestamp>",
```
Comment (brandond) [inline pkg/cli/cmds/etcd_snapshot.go:31]:
```suggestion
				Usage:       "(db) Set the base name of on-demand etcd snapshots. Default: on-demand-",
```

File: pkg/cli/cmds/etcd_snapshot.go [hunk 3]
```diff
@@ -24,19 +27,14 @@ func NewEtcdSnapshotCommand(action func(*cli.Context) error) cli.Command {
...
 				Destination: &ServerConfig.EtcdSnapshotName,
-			},
-			&cli.IntFlag{
-				Name:        "etcd-snapshot-retention",
-				Usage:       "(db) Number of snapshots to retain",
-				Destination: &ServerConfig.EtcdSnapshotRetention,
-				Value:       defaultSnapshotRentention,
+				Value:       "on-demand-" + strconv.Itoa(int(time.Now().Unix())),
```
Comment (brandond) [inline pkg/cli/cmds/etcd_snapshot.go:33]: The current timestamp still gets appended to this later, correct?

```suggestion
				Value:       "on-demand-",
```

**Review by brandond — CHANGES_REQUESTED** [review]
Minor nits to the name/etc-snapshot-name UX.

Note that you don't have to explicitly put the default in the usage text if you specify a value; the CLI library does this for you:
```
[root@centos01 ~]# k3s etcd-snapshot --help | grep name
   --name value (db) Set the name of the etcd on-demand snapshot. Default: on-demand-<unix-timestamp> (default: "on-demand-1611166541")
[root@centos01 ~]# k3s etcd-snapshot --help | grep name
   --name value (db) Set the name of the etcd on-demand snapshot. Default: on-demand-<unix-timestamp> (default: "on-demand-1611166542")
[root@centos01 ~]# k3s etcd-snapshot --help | grep name
   --name value (db) Set the name of the etcd on-demand snapshot. Default: on-demand-<unix-timestamp> (default: "on-demand-1611166543")
```

File: pkg/cli/cmds/etcd_snapshot.go (hunk 3 above)
Comment (briandowns) [reply pkg/cli/cmds/etcd_snapshot.go:33]: No, which I think relates to the comments above. Timestamp is only used as a suffix when no name is provided. I'll need to  rework some things to have it appended regardless.

File: pkg/etcd/etcd.go [hunk 4]
```diff
@@ -712,48 +710,80 @@ func snapshotDir(config *config.Control) (string, error) {
...
 
 	cfg, err := getClientConfig(ctx, e.runtime, endpoint)
 	if err != nil {
-		logrus.Errorf("Failed to get config for etcd snapshot: %v", err)
-		return
+		return errors.Wrap(err, "failed to get config for etcd snapshot")
 	}
 
-	snapshotPath := filepath.Join(snapshotDir, snapshotPrefix+strconv.Itoa(int(time.Now().Unix())))
+	snapshotName := e.config.EtcdSnapshotName + "-" + strconv.Itoa(int(time.Now().Unix()))
```
Comment (brandond) [inline pkg/etcd/etcd.go:761]:
```suggestion
	snapshotName := fmt.Sprintf("%s-%d", e.config.EtcdSnapshotName, time.Now().Unix())
```

File: pkg/etcd/etcd.go [hunk 5]
```diff
@@ -712,48 +710,80 @@ func snapshotDir(config *config.Control) (string, error) {
 	return config.EtcdSnapshotDir, nil
 }
 
-// snapshot attempts to save a new snapshot to the configured directory, and then clean up any old
-// snapshots in excess of the retention limits.
-func (e *ETCD) snapshot(ctx context.Context) {
+// preSnapshotSetup checks to see if the necessary components are in place
+// to perform an Etcd snapshot. This is necessasry primarily for on-demand
```
Comment (brandond) [inline pkg/etcd/etcd.go:714]:
```suggestion
// to perform an Etcd snapshot. This is necessary primarily for on-demand
```

**Review by brandond — COMMENTED** [review]
LGTM except for one docstring typo I just noticed

**Review by brandond — APPROVED** [review]

**Comment by galal-hussein** [comment]
The PR looks good to me, the only comment I have is that I can see users confuse the command with non-embedded etcd k3s, meaning that when user trigger this command with kine it may return etcd error saying that it cant connect to etcd or sometthing similar, it may benefit us to display an error stating that etcd embedded isnt set up.

**Comment by briandowns** [comment]
If this is ran on a system not running etcd, this is the output:

```sh
FATA[2021-01-20T20:25:37.727386821Z] managed etcd database has not been initialized 
```

**Comment by galal-hussein** [comment]
> If this is ran on a system not running etcd, this is the output:
> 
> ```shell
> FATA[2021-01-20T20:25:37.727386821Z] managed etcd database has not been initialized 
> ```

thats good enough for me, LGTM

**Review by galal-hussein — APPROVED** [review]
LGTM

**Pull request merged**